### ISSUE TITLE
DEP: linalg: deprecate `kron`

### ISSUE THREAD

**Issue by ilayn** [scipy/scipy#20077]
`linalg.kron` is a basic, data-movement intensive array operation. There is hardly any computation done other than scalar times array and quite some shuffling needed for higher order nd arrays. It is similar to a very special case of `np.einsum` in higher dimensions. Hence it doesn't make too much sense for us to carry around a suboptimal version of it since Numpy already supports batch array input https://numpy.org/doc/stable/reference/generated/numpy.kron.html

For comparison we also have Khatri-Rao product which is a Kronecker-like product and it is specific enough for us to have it but kron is not in my opinion.

See #20072 and https://github.com/pymc-devs/pytensor/issues/640 for more context.

**Comment by ev-br** [comment]
Why do we have scipy.linalg.kron at all? Burn it fire, I'd say.

**Comment by j-bowhay** [comment]
I had been meaning to propose this for a while as the numpy version also supports higher dimensions.

**Comment by ilayn** [comment]
Then should I leave it to you @j-bowhay ? I'm not sure how the deprecation tracker works so no problem for me if you take over.

**Comment by j-bowhay** [comment]
> Then should I leave it to you @j-bowhay ? I'm not sure how the deprecation tracker works so no problem for me if you take over.

Yes I am happy to tackle this

**Comment by rgommers** [comment]
This does not look like the right approach to me, for a few reasons:

1. our `kron` isn't broken, just doesn't have as many features as the numpy version of it. It's also not in the wrong place or has a bad API. So in principle, the deprecation is just churn for users.
2. `scipy.linalg` is kinda designed to be a large superset of `numpy.linalg`. And in principle can have higher performance (although that's not relevant here). So users who have SciPy as a dependency should be able to use just `scipy.linalg` rather than mix and match depending on the function
3. we may want the SciPy versions to be generic over array types in the end, meaning our `kron` may end up providing a superset of functionality compared to the numpy version of it

Signatures look compatible, so why not simply alias our `kron` to `numpy.linalg.kron` for now? That's less disruptive for users, easier to do that a deprecation trajectory, and keeps goals (2) and (3) alive.

Minor procedural side note: please don't close deprecation proposal issues within a day before they're addressed.

**Comment by j-bowhay** [comment]
This also seems fine to me but I think we should then consider the linalg roadmap item "Reduce duplication of functions with numpy.linalg" which seems contradictory to 2). My initial support for this was under the premise of working towards this item.

**Comment by ilayn** [comment]
This functionality is hardly a linalg thing. So I don't agree that it has to stay here. We need to separate array manipulation functions from actual linalg stuff. Transpose is also an array op but has no place in scipy to have its own copy here. This distinction will help us to weed out algorithms from memory ops in the long run in my opinion. 

> please don't close deprecation proposal issues within a day before they're addressed.

This was more or less after folks agreed upon it.

**Comment by ev-br** [comment]
I think this issue is gives a good point to step back and ask: what actually is `scipy.linalg`?

1. The historic description was along the lines of "provides LAPACK-accelerated versions of numpy.linalg functions". Is it still relevant, is numpy using lapack-lite?
2. A home for additional/more advanced linalg primitives (qr_update etc)
3. (new) An umbrella dispatcher to various backends (numpy, cupy, pytorch, jax etc) via Array API.

I sort of sense a potential tension between 2  and 3, but that's separate from `kron`.

Now, what is `scipy.linalg.kron`? It shares the name with a `numpy.linalg` function, but is not an accelerated or otherwise improved version of it. 

Does it provide something new and useful? I don't see much --- a fourth slightly incompatible version of `multiply.outer`, `np.linalg.kron` and `scipy.linalg.khatri_rao` does not sound too convincing IMO.

Can it be replicated on a user side? Yes, easily, the whole implementation is three LOC: https://github.com/scipy/scipy/blob/v1.12.0/scipy/linalg/_special_matrices.py#L448-L486

So, yes, deprecating it is backcompat break. However I think it's a good change, especially if made together with some fleshing out of what where we want `scipy.linalg` to evolve.

**Comment by ilayn** [comment]
These days I have been guilty of typing things a bit too terse and sound a bit pedestrian, so let me flesh out sentences a bit what I meant. 

Even though numpy and scipy has the namespace identical is probably a bit unfortunate because that gives the impression that NumPy and SciPy has the same goal but with different scope when it comes to linalg. It might also be explained as "not as comprehensive as scipy but better than nothing". I don't share this view but I can see how it can be articulated. 

In my opinion NumPy is and should be the place that provides best-of-class array manipulations and memory tricks and SIMD and so on. If the operation has a mathy name, this should not be a block or denounce it from adding to NumPy. 

Now the question is, how would the user know whether it is a memory trick or a real full algorithm like LU or SVD. That is, I think, a very valid point and I don't have a conclusive answer to. However borrowing NumPy funcs into the scipy namespace with an alias does not seem like a good long term solution. We are removing them one by one almost exclusively so far. 

In this particular case, we are actually a bit lucky because it lives under `numpy` namespace and not in `numpy.linalg` hence the confusion is less dramatic. 

Like I mentioned above, array tricks like `tile`, `fill` or `concatenate` and other things are rightly inside NumPy and I think they should be there. In my limited linalg view `kron` is just like tile but has a mathy name for it. Hence probably should better live in NumPy namespace.

**Comment by lucascolley** [comment]
It's difficult to weigh up the balance of the code-churn (which is clearly 'unnecessary' in at least some reasonable sense), and trying to slim `scipy.linalg` down (given that the duplication between SciPy and NumPy is definitely a point of confusion for users).

Without my array API hat on, I think the arguments given for why `linalg.kron` does _not_ belong in SciPy are strong. I think the code-churn could be justified by reducing confusion and improving maintainability long-term.

---

_However_, I think Evgeni's 3rd point is important. Sure, it is only 3 LOC, but not completely trivial, and (probably?) we will want a (somewhat) 'standard' array-agnostic implementation to live somewhere long-term. If that isn't SciPy, where else?

1. The array API standard (main namespace? unlikely. `linalg` extension? perhaps?)
2. A different existing package (which?)
3. A new package, specifically for building array-agnostic functions _on top of_ the standard. Perhaps targeted at things between the array library and SciPy levels (e.g. `xp.cov`), which could then be candidates for additions to the standard.

If there are concrete plans for any of options 1-3, then I'd probably be slightly in favour of the deprecation.

However, if there is a risk that we will deprecate, but then in future wish to _return_ `kron` to `scipy.linalg` (albeit serving a slightly different purpose), that would almost certainly be a bad and confusing amount of code-churn. What do we think?

**Comment by ilayn** [comment]
> If that isn't SciPy, where else?

Why not NumPy? It is already there. We are not removing from NumPy too. Only from `scipy.linalg`.

> However, if there is a risk that we will deprecate, but then in future wish to return kron to scipy.linalg (albeit serving a slightly different purpose), that would almost certainly be a bad and confusing amount of code-churn. What do we think?

I don't see how a kron operator would gain a different purpose. It has a very well defined operator definition just like `numpy.outer`. I would have made the same proposal if numpy had [block_diag](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.block_diag.html) but unfortunately it doesn't so scipy.linalg it is. Though it should have been NumPy all along.

**Comment by lucascolley** [comment]
Sorry, probably best to avoid a rapid back-and-forth here but just to clarify:

> Why not NumPy?

I'm talking specifically about the function being available regardless of which array library is used. Perfectly reasonable to say that every library should have it, but in that case we'd want a plan for it to be added to the standard.

> I don't see how a kron operator would gain a different purpose.

Sorry, I phrased this poorly. I meant just the ability to take any array library as input, not just NumPy.

**Comment by ilayn** [comment]
>  but in that case we'd want a plan for it to be added to the standard.

I think I don't follow this part. It is already in NumPy namespace. Why should we the arbiter of what array api standard should have? We have not been active in the `linalg` parts of it so far. 

In this case, we are removing things that are not suitable for us. What happens afterwards are the packages' prerogative as not every NumPy function is available in the other namespaces anyways.

**Comment by rgommers** [comment]
> I think this issue is gives a good point to step back and ask: what actually is `scipy.linalg`?

Yes, that is a useful exercise indeed. Probably better in a fresh issue? This one is very specific.

> So, yes, deprecating it is a backcompat break. However I think it's a good change, especially if made together with some fleshing out of where we want `scipy.linalg` to evolve.

First deprecating and then thinking about the bigger picture and whether we need it seems clearly the wrong way around to me. Why not just make it an alias right now, and if someone actually has energy to work on defining the plan for `scipy.linalg` and pushing it forward, then it can either be deprecated or not.

**Comment by rgommers** [comment]
> I think I don't follow this part. It is already in NumPy namespace

Please let's not do another round of this discussion right now? We've already been through this I'm sure. The tl;dr is: numpy is an array library, it will not be deferring to other array libraries. SciPy can support multiple array libraries.

**Comment by ilayn** [comment]
OK I am definitely not following. How is array api relevant here? I am not even trying to start any discussion let alone that one in particular. It is already confusing enough. This function is not in the standard plus as I argued above it is not even suitable for us.

**Comment by lucascolley** [comment]
There seems to be agreement that a `kron` which works only for NumPy arrays belongs in NumPy.

---

The question is where a `kron` which works with _any_ array library belongs.

This may well not be SciPy - for many of the same reasons given in this thread for why it belongs on (or at least closer to) the array library level. When considering the following options:

1. In the array libraries (i.e. `xp.kron` or `xp.linalg.kron`, in the standard),
2. In SciPy, or
3. In a different library (see my point (3) above),

my preference would be for option 1 or 3.

_However_, we must acknowledge that option 2 is the closest to reality. Such a library as option 3 does not exist, and additions to the standard (and their implementations) are an involved process. Either one involves work. Option 2 would simply be a ~10 line PR.

---

The point is _not_ that SciPy should be the arbiter of what belongs in the standard, nor that SciPy should be responsible for providing implementations that would otherwise be out of scope. Rather, at a time where we are asking "what actually is `scipy.linalg`?", support for multiple array libraries is a part of the conversation.

I don't think anyone has made an argument for `scipy.linalg.kron` staying long-term. However, perhaps it would be best to figure out a plan based on the question "what actually is `scipy.linalg`?" (_in a separate issue_), and how (if at all) array-agnostic functions like this fit into it (in the short- or long-term), before we start causing code-churn. If we don't want functions like that in SciPy, it would be nice to find another home for them[^1] first. Finding that home is not SciPy's responsibility, but we might be doing whoever puts in that work a favour by being patient.

[^1]: The reason discussion of this is difficult to get right is the difference between implementation and API. While we have already found another home for the implementation (NumPy), the API for an array-agnostic implementation does not have a home yet. SciPy's existing `kron` API is the most immediate (though not certainly the best) candidate for a home.

**Comment by ilayn** [comment]
Yes that's perfectly fine and I get the argument. The argument I am failing to convey is that kron does not have to be in every package. This is not a core linalg functionality. 

If we remove it, we are not failing anyone so why can't we just deprecate it is still unanswered. That's why I don't see the connection why we are considering all these 3 options. Just like we removed tril and triu we can remove this one too. So I need some clarification why kron is more important than those two.

**Comment by rkern** [comment]
`tril` and `triu` are in the Array API. `kron` is not. There is an opportunity for `scipy` to provide an Array API-agnostic, fully-featured `kron` implementation so no one else has to.

**Comment by rkern** [comment]
We do use `kron` elsewhere in `scipy`, so to make those parts of `scipy` Array API-agnostic, we'd need an Array API-agnostic version of `kron` anyways, so to move that forward, we wouldn't want to rely on using `np.kron`.

**Comment by ilayn** [comment]
Ok let me try another way, kron was until now never meant to be in the array api right? So just because historically we had a bad version of it should not make it eligible to an addition to array api. That doesn't make any sense. I really don't know how else I can put this. 

We are also using extensively triu and tril **and** concatenate and tile and many others that are not ever considered for array api. No argument is done on them because those also equally would make no sense. Because there is no historical luggage on them. This one has and I am strongly -1 on making it our responsibility

**Comment by ilayn** [comment]
Put differently this function has no place in scipy.linalg because it is not a linalg function. It is an array creation function.

**Comment by rkern** [comment]
[`tril`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.tril.html) and [`triu`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.triu.html) _are_ in the Array API. Therefore, `scipy` does not need to provide copies of them because we can rely on the Array API implementation to provide them.

No one is proposing to add `kron` to the Array API. However, `scipy` does use `kron` in various places. If we want to make those places Array API-_agnostic_ (not implementations of the Array API but just code that can work with whatever Array API implementation objects it is given), then we need an Array API-agnostic version of `kron`. To the extent that `np.kron()` will not preserve the Array API implementation of its inputs and instead convert them to `np.ndarray`s and return an `np.ndarray`, we need an implementation that does preserve the implementation of the inputs. This is not the same as adding `kron` to the Array API. To the contrary, if it _were_ added to the Array API, then we could easily go forward with the deprecation of `scipy.linalg.kron`.

> > we wouldn't want to rely on using np.kron

> Quite the other way, we would like to solely depend on np.kron

Please be careful with snipping out phrases. I start with "to move that [Array API agnosticization] effort forward, we wouldn't ...". And that is true. If you rely on `np.kron`, that's a blocker for moving that effort forward. If you don't consider that effort, then yes, the balance would favor relying on `np.kron`. This issue started without considering that effort, and was reopened in part to consider that effort.

> Put differently this function has no place in scipy.linalg because it is not a linalg function. It is an array creation function.

It's a special matrix. We have a [whole module in `scipy.linalg` for creating such special matrices](https://github.com/scipy/scipy/blob/main/scipy/linalg/_special_matrices.py).

**Comment by ilayn** [comment]
OK I am about to give up. First thing, we already agreed initially this function does not make sense here. 

Then Ralf comes in he says no, that's a bit out of place but fine. Then we start discussing this function should support array api. 

Let's focus on this instance. Forget about anything comes after this. This function should not support array API because it doesn't have a place in scipy.linalg. 

1- it is not a special matrix. There is no such thing called Kronecker matrix. This is a misnomer. 
2- it is an array creation routine and it is already in the numpy namespace for this particular reason. You supply arrays and you generate a new one.

On top of it this function is very slow if you do it on Python level because you need to allocate a new array, place the existing inputs by scalar multiples. Just like einsum this goes much faster if there is a numpy implementation done on the C level. Anything else is just a historical convenience and should not have been there. 

If I can't make this point across I guess I have to concede this argument however I have to confess so far array api or counter arguments for not deprecating is not put forward and still stuck on array api.

**Comment by lucascolley** [comment]
In principle, a `kron` function which works when given a CuPy array, a Dask array or a NumPy array (etc.) is desirable to _use_ in many projects. SciPy happens to be one of these projects, as Robert pointed out. It can live anywhere, as long as a project like SciPy can use it. Point taken about a Python implementation being slow, but the ability to handle various array types holds value in spite of performance concerns[^1].

You have made clear that you believe that this function[^2] doesn't belong in `scipy.linalg`. I'm inclined to agree with you. However, there are no obvious other places for it right now. Given that it is plausible (if undesirable) that this will find its home (perhaps temporarily) in `scipy.linalg`, it would be awkward to deprecate now. Hence my suggestion to go ahead with the deprecation of the current `scipy.linalg.kron`, but only once we can be sure that this array-agnostic function will live somewhere else, so that we (and other libraries) can use it. Since that requires work from others, we can be patient, alias `np.kron`, and live with the unfortunate naming for a little longer.

[^1]: In SciPy we could use `np.kron` to be performant with NumPy arrays, and use this 'array-agnostic' function for other types.

[^2]: To be crystal clear, 'this function' refers to the `kron` that works with any array API compatible array, throughout.

**Comment by ilayn** [comment]
> However, there are no obvious other places for it right now

That's circling the argument. If a project needs it, they can implement it in their native code. It does not need a common place. Let's not go back to the beginning again. 

Why do you want it to work with different arrays, could you elaborate? If a function drops out of SciPy why do we need to find a new home for it?

**Comment by j-bowhay** [comment]
> Why do you want it to work with different arrays, could you elaborate? If a function drops out of SciPy why do we need to find a new home for it?

We use `kron` within `solve_discrete_lyapunov`. `solve_discrete_lyapunov` is not in the standard so we would have to provide an implementation that works will other array rather than just dispatching to a library's implementation. Hence as `kron` is used within `solve_discrete_lyapunov` and is also not in the standard we also would need an implementation of `kron` that works with other array types than just NumPy.

**Comment by ilayn** [comment]
Yes exactly, so then `kron` has to go into the standard if solve_are or solve_lyapunov has to support the standard. Or it has to fall back to NumPy arrays. 

We discussed this and just as above I would really prefer not to start that discussion again. Every function in SciPy cannot support array API because otherwise SciPy is shackled by array API forever and cannot move since every internal function it uses has to support array API. That is exactly what we have established that it won't happen in the first place when we started doing this. Instead it has to add piecemeal support for things that it can support.

**Comment by lucascolley** [comment]
> If a array library needs it, they can implement it in their native code. 99% of the users of kron will switch to NumPy. It does not need a common place. 

I am not talking about end-user code. I am talking about consumer libraries like SciPy, which have the potential to become array-agnostic, that currently use `np.kron` or `torch.kron` or `scipy.linalg.kron` or `cupy.kron` etc. Rather than every one of these consumer projects providing its own new `kron` implementation[^1], it probably makes sense to have a single implementation which everyone can use. I'm _not_ arguing for the view that this single implementation _must_ exist_, nor that it _must exist in SciPy_. I'm arguing for the view that it may be helpful to the ecosystem for this single implementation to exist, and SciPy is plausibly at least a temporary home for it.

[^1]: Of course this would be possible given that it is three lines of code, but having to move away from a single function to three nontrivial lines many times over is still not ideal.

**Comment by ilayn** [comment]
@lucascolley that is still again circling back to the starting point, we already, I hope, established that this is an array function. So it cannot be array agnostic. That is the reason why we want to deprecate it in the first place. 

See my comment above. We cannot maintain a library if there is an external dependency that blocks us by its own state.

Can we stop talking about array api for this please now? I don't know how different can I make the same point that array api does not apply to this function.

**Comment by rgommers** [comment]
> we already, I hope, established that this is an array function. So it cannot be array agnostic.

I don't think we've established that, since no one commented on the implementation beyond simply asserting this. And it does not seem true. If you look at the numpy implementation, then you'll see (modulo some `np.matrix`-related cruft that we don't need) that it is implemented in pure Python and can be done so with the primitives: `asarray`, `max`, `reshape`, `expand_dims` and `+`/`=`. So it looks to me like we are able to write a `kron` that's generic over array types, while being as performant as the numpy version.

How about I just write that implementation? That seems like a more effective use of time than more comments here.

**Comment by ilayn** [comment]
I would really need to understand why we are trying to keep kron here. What is the concrete downside other than a hypothetical array agnosticism (which is currently already not there)?

Unless you want to write C code for it, no matter what you write here won't make any difference in terms of performance as this has to happen in the low level if we want a decent implementation. Here are some random data points from the interwebz why this is fundamentally a memory manipulation function

https://stackoverflow.com/questions/60383660/why-is-numpys-kron-so-fast
https://discourse.julialang.org/t/kron-vs-scalar-product-speed-difference-python-code-faster/1460/42
https://nl.mathworks.com/matlabcentral/answers/111918-is-it-possible-to-improve-performance-of-the-kron-operator-by-making-it-a-built-in-function-rather-t

in fact first one shows happens to demonstrate the einsum connection I mentioned above. I hope we are not going to take over einsum too. So concrete question, why are we so willing to commit to maintaining a less performant version of kron? Array libraries can do this much more conveniently in their own native format be it distributed or otherwise.

**Comment by fancidev** [comment]
IIUC, Party A wants to deprecate `scipy.linalg.kron` to keep the API surface neat. Party B wants to keep it to work with different array back-ends.

For _library_ writers targeting multiple back ends (e.g. Party B), a function like this is _necessary_ after all. On the other hand, Party A thinks such a function is not the business of SciPy’s (public) API.

So it appears to me that a private API, such as `scipy.linalg._kron` that does the job, should work with both parties to start?

As for whether to deprecate (and later remove) the public API, it won’t affect me (a light numpy user) as I could use `numpy.kron` instead.

**Comment by mdhaber** [comment]
> So it looks to me like we are able to write a kron that's generic over array types, while being as performant as the numpy version. How about I just write that implementation? That seems like a more effective use of time than more comments here.

I had about the same thought, wrote the function, and then came across this comment. Here it is:

<details>
For lines that have changed (other than those that use `xp_expand_dims` instead of `np.expand_dims`), the original NumPy code is included as a comment. 

```python3
import numpy as np
import numpy
import array_api_strict
import jax.numpy
import torch
# import cupy

from scipy._lib._array_api import array_namespace

def xp_expand_dims(a, axis):
    xp = array_namespace(a)
    for i in axis:
        a = xp.expand_dims(a, axis=i)
    return a

def kron(a, b):
    xp = array_namespace(a, b)
    b = xp.asarray(b)  # b = asanyarray(b)
    singletons = (1,) * (b.ndim - a.ndim)
    a = xp.broadcast_to(xp.asarray(a), singletons + a.shape)  # a = array(a, copy=None, subok=True, ndmin=b.ndim)
    # is_any_mat = isinstance(a, matrix) or isinstance(b, matrix)
    ndb, nda = b.ndim, a.ndim
    nd = max(ndb, nda)
    if (nda == 0 or ndb == 0):
        return xp.multiply(a, b)  # _nx.multiply(a, b)

    as_ = a.shape
    bs = b.shape
    # if not a.flags.contiguous:
    #     a = reshape(a, as_)
    # if not b.flags.contiguous:
    #     b = reshape(b, bs)

    # Equalise the shapes by prepending smaller one with 1s
    as_ = (1,)*max(0, ndb-nda) + as_
    bs = (1,)*max(0, nda-ndb) + bs

    # Insert empty dimensions
    a_arr = xp_expand_dims(a, axis=tuple(range(ndb-nda)))
    b_arr = xp_expand_dims(b, axis=tuple(range(nda-ndb)))

    # Compute the product
    a_arr = xp_expand_dims(a_arr, axis=tuple(range(1, nd*2, 2)))
    b_arr = xp_expand_dims(b_arr, axis=tuple(range(0, nd*2, 2)))
    # # In case of `mat`, convert result to `array`
    result = xp.multiply(a_arr, b_arr)  # result = _nx.multiply(a_arr, b_arr, subok=(not is_any_mat))

    # Reshape back
    # yes, use np.multiply on the shapes `as_` and `bs_`
    result = xp.reshape(result, tuple(np.multiply(as_, bs)))  # result.reshape(_nx.multiply(as_, bs))

    return result  # if not is_any_mat else matrix(result, copy=False)

# example from NumPy documentation
for xp in [numpy, array_api_strict, jax.numpy, torch]:
    a = xp.reshape(xp.arange(100), (2,5,2,5))
    b = xp.reshape(xp.arange(24), (2,3,4))
    c = kron(a,b)
    assert c.shape == (2, 10, 6, 20)
    assert type(c) is type(xp.asarray(1))
    I = (1, 3, 0, 2)
    J = (0, 2, 1)
    J1 = (0,) + J             # extend to ndim=4
    S1 = (1,) + b.shape
    K = tuple(xp.asarray(I) * xp.asarray(S1) + xp.asarray(J1))
    assert c[K] == a[I]*b[J]
```

</details>

For small arrays, the array API calls introduce some overhead, but for large arrays, the performance of this implementation with NumPy arrays is the same as `np.kron`. I'd be happy to submit the PR if there's someone to review it (presumably by comparing against the NumPy documentation and code).

**Comment by lucascolley** [comment]
I can review!

**Comment by lucascolley** [comment]
Thanks for the benchmarks Ilhan. Indeed, this is evidence that we would like to see `kron` in the standard eventually, for efficiency's sake.

Until that happens, merging Matt's implementation seems pragmatic while we are unsure about the future landscape. If we get a C-API version of `np.kron` or a Cython version of `scipy.linalg.kron` merged, we can accommodate that via

```python
xp = array_namespace(x)
if is_numpy(xp):
    # delegate to the fast implementation, wherever it lives
else:
    # use the Python level implementation for other backends
```

That leaves open the option of delegating to other libraries or using a standardised API in the future.

**Comment by ilayn** [comment]
The proposal is to deprecate it out of SciPy and use `xp.kron`/`np.kron` wherever needed just like we did with `triu`/`tril`.

**Comment by lucascolley** [comment]
> The proposal is to deprecate it out of SciPy and use xp.kron/np.kron wherever needed just like we did with triu/tril.

I'm happy to help with this (if/)when `xp.kron` exists in the standard, like `xp.triu` and `xp.tril`.

**Comment by mdhaber** [comment]
Those benchmarks look as expected @ilayn. I wouldn't claim that we can beat either NumPy or SciPy with a direct array API translation because of the extra overhead (e.g. going through `array_api_compat`, looping over `expand _dims`). The overhead for small problems was actually much worse on my machine.

Honestly, the motivation here is to minimize the time spent thinking about `kron`. If we include this, "dispatching" to array library native implementations when available and possibly special-casing to a translation of the SciPy code for cases in which it would be faster, we can have a version that is about as fast as the best available, if it matters. And it would close a few issues.

I hesitate to open a PR against your continued objection, though! Let me know if you'd like to be done with these and I'll open the PR or if you'd prefer to continue the discussion and I'll move on.

**Comment by ilayn** [comment]
The original issue was to ask whether we want to deprecate or not but somehow it is getting to a different implementation. And somehow Array API became the central argument point which after all these I still have no idea why it is even relevant. It should not be dispatched to anything but the array API libraries should offer this. NumPy already does it. The rest does not interest me at this point.

My objection is to keep it for the sake of Array API which should not play a role in this discussion. Somehow there is a significant support for this Array API that I have no use case for and not sure whether I should really care about. So far I did not understand why Array API should be relevant for this function and in fact why should be a part of any discussion about SciPy (I don't have any capacity left to reignite it). 

The benchmark above is to show that there is a significant performance loss by keeping it here. This is an array creation function with using two array as the input data. And the array generation must be done at the source of these arrays NumPy or Dask or anything else. I am running out of different ways to spell this out after all this discussion. 

So I guess the answer is no we don't deprecate it here and accept the performance loss as the conclusion here.

**Comment by lucascolley** [comment]
> accept the performance loss as the conclusion here

If a faster implementation is merged to `np.kron`, we can just delegate to that - why would that mean accepting performance loss? I think the performance considerations are orthogonal to deprecation, unless I am missing something.

**Comment by rkern** [comment]
I'd go ahead with deprecating `scipy.linalg.kron` in favor of `np.kron`.

When we run into the task of Array API-agnosticizing scipy functions that use `kron`, we can add Matt's implementation (with `np`-delegation) to `array_api_compat` and use it from there, and submit `kron` as a proposed addition to the next version of the Array API for fast native implementations.

**Comment by lucascolley** [comment]
I'll open an issue on the array API repo now 👍 we are really playing guessing games here until the work is put in on the array library side.

> If I won't offend anyone with my waiver from this discussion, I really want to get this Fortran stuff out the door in my limited capacity.
> 
> Lately, I am getting a bit overwhelmed by trying to make my point across and this is probably my own fault due to my lack of clear argumentation and very likely that it comes across as dismissive. So any decision OK for me regarding this issue. I hope you don't find it as disrespectful or cynical.

Absolutely no worries at all from my side - this discussion was useful to work through as we may find ourselves in a similar situation with other functions. Good luck with the Fortran!

**Comment by lucascolley** [comment]
to close out here, an array-agnostic `kron` should be available with the next release of array-api-extra: https://data-apis.org/array-api-extra/generated/array_api_extra.kron.html
